utils/zerovec/src/varzerovec/mod.rs

```diff
@@ -0,0 +1,417 @@
+use std::ops::Index;
+
+mod components;
+#[cfg(feature = "serde")]
+mod serde;
+
+/// A zero-copy vector for variable-width types.
+///
+/// `VarZeroVec<T>` is designed as a drop-in replacement for `Vec<T>` in situations where it is
+/// desirable to borrow data from an unaligned byte slice, such as zero-copy deserialization, and
+/// where `T`'s data is variable-length (e.g. `String`)
+///
+/// `T` must implement [`AsVarULE`], which is already implemented for [`String`]. References are obtained
+/// as its [`AsVarULE::VarULE`] type, which in the case of [`String`] is [`str`].
+///
+/// `VarZeroVec<T>` behaves much like [`std::borrow::Cow`], where it can be constructed from owned data
+/// but can also borrow from some buffer.
+///
+/// # How it Works
+///
+/// `VarZeroVec<T>`, when used with non-human-readable serializers (like `bincode`), will
+/// serialize to a specially formatted list of bytes. The format is:
+///
+/// - 4 bytes for `length` (interpreted as a little-endian u64)
+/// - `4 * length` bytes of `indices` (interpreted as little-endian u64s)
+/// - Remaining bytes for actual `data`
+///
+/// Each element in the `indices` array points to the starting index of its corresponding
+/// data part in the `data` list. The ending index can be calculated from the starting index
+/// of the next element (or the length of the slice if dealing with the last element).
+///
+/// # Safety
+///
+/// `VarZeroVec<T>` is implemented with a fair amount of unsafe code, but is externally
+/// safe to use.
+///
+/// # Example
+///
+/// ```
+/// # use std::str::Utf8Error;
+/// # use zerovec::VarZeroVecError;
+/// use zerovec::VarZeroVec;
+///
+/// // The little-endian bytes correspond to the list of strings.
+/// let strings = vec!["w".to_owned(), "ω".to_owned(), "文".to_owned(), "𑄃".to_owned()];
+/// let bytes = &[
+///     4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0,
+///     0, 0, 6, 0, 0, 0, 0, 0, 0, 0, 119, 207, 137, 230, 150, 135, 240, 145, 132, 131,
+/// ];
+///
+/// let zerovec: VarZeroVec<String> = VarZeroVec::try_from_bytes(bytes)?;
+///
+/// assert_eq!(zerovec.get(2), Some("文"));
+/// assert_eq!(zerovec, &*strings);
+/// # Ok::<(), VarZeroVecError<Utf8Error>>(())
+/// ```
+///
+/// [`ule`]: crate::ule
+#[derive(Clone)]
+pub struct VarZeroVec<'a, T>(VarZeroVecInner<'a, T>);
+
+/// Implementation details of VarZeroVec
+///
+/// Should not be made public.
+///
+/// Safety invariant: Borrowed() must have the appropriate
+/// format:
+///
+/// The format of the slice is conceptually:
+/// ```rust,ignore
+/// struct SliceFormat {
+///     len: u64::ULE,
+///     lens: [u64::ULE],
+///     things: [u8]
+/// }
+/// ```
+///
+/// `len` is the length of `lens`, and each element of `lens` is the starting
+/// index of the thing in `things`.
+///
+/// The actual implementation details of this can be found in the `components` module
+#[derive(Clone)]
+enum VarZeroVecInner<'a, T> {
+    Owned(Vec<T>),
```

Comment:
Suggestion: In ZeroVec, I made the Owned variant have the same memory layout as the Borrowed variant.  This leads to simpler code in this wrapper class.  Can we make Owned be `Vec<u8>` with the correct byte layout?

Reply:
This choice is intentional, it allows for `VarZeroVec` to be mutated easily to allow for "patching". This kind of patching is way harder with our `Vec<u8>` format because each operation basically requires a new allocation and a full vector copy.